llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -3421,6 +3421,29 @@ Value *InstCombinerImpl::foldAndOrOfICmps(ICmpInst *LHS, ICmpInst *RHS,
   return foldAndOrOfICmpsUsingRanges(LHS, RHS, IsAnd);
 }
 
+Value *foldAorBZero(BinaryOperator &I, InstCombiner::BuilderTy &Builder) {
+  Value *Op0 = I.getOperand(0);
+  Value *Op1 = I.getOperand(1);
+  if (!Op0->hasOneUse() || !Op1->hasOneUse())
+    return nullptr;
+
+  // match each operand of I with and
+  Value *A, *B;
+  CmpInst::Predicate Pred = CmpInst::ICMP_EQ;
+  CmpInst::Predicate InPred = CmpInst::ICMP_EQ;
+  bool IsOp0 = match(Op0, m_c_And(m_Cmp(Pred, m_Value(A), m_ZeroInt()),
+                                  m_Cmp(InPred, m_Value(B), m_ZeroInt())));
+  bool IsOp1 = match(Op1, m_c_And(m_Cmp(InPred, m_Specific(A), m_ZeroInt()),
+                                  m_Cmp(Pred, m_Specific(B), m_ZeroInt())));
```

Comment:
Shouldn't this fold also work for any other value? 

I am not really familiar with llvm internals, but looks like code is specific to 0 constant

Reply:
Yeah, I think the current implementation is too specific. It also works for pointer compares `(a == nullptr && b != nullptr) || (a != nullptr && b == nullptr)`.

We can reuse the `isKnownInversion` helper here.

Reply:
I can't find `isKnownInversion` any where being used do you have examples for it? 

I  tried checking on alive to see a compare against any value and it works. 
https://alive2.llvm.org/ce/z/C7Tann

Reply:
See https://github.com/llvm/llvm-project/pull/93591